platforms/jvm/testing-jvm/src/integTest/groovy/org/gradle/testing/nonclassbased/IncorrectSetupNonClassBasedTestingIntegrationTest.groovy

```diff
@@ -0,0 +1,233 @@
+                            testDefinitionDirs.from("$DEFAULT_DEFINITIONS_LOCATION")
+                        }
+                    }
+                }
+            }
+        """
+
+        // Ensure test definitions dir exists even if no defs are added
+        file("$DEFAULT_DEFINITIONS_LOCATION").createDir()
+
+        if (hasTestClasses) {
+            writeTestClasses()
+        }
+        if (hasTestDefs) {
+            writeTestDefinitions()
+        }
+
+        when:
+        if (shouldFail) {
+            fails("test", "--info")
+        } else {
+            succeeds("test", "--info")
+        }
+
+        then:
+        if (shouldBeSkipped) {
+            testTaskWasSkippedDueToNoSources()
+        } else if (shouldFail) {
+            sourcesPresentAndNoTestsFound()
+        } else {
+            if (scanForTestClasses && hasTestClasses) {
+                classBasedTestsExecuted()
+            }
+            if (addTestDefsDir && hasTestDefs) {
+                nonClassBasedTestsExecuted()
+            }
+        }
+
+        where:
+        scanForTestClasses  | hasTestClasses    | addTestDefsDir | hasTestDefs || shouldBeSkipped || shouldFail
+        true                | true              | true              | true          || false            || false
+        true                | false             | true              | true          || false            || false
+        true                | true              | false             | true          || false            || false
+        true                | true              | true              | false         || false            || false
+        true                | false             | false             | true          || true             || false
+        true                | false             | true              | false         || true             || false
+        true                | true              | false             | false         || false            || false
+        true                | false             | false             | false         || true             || false
+        false               | true              | true              | true          || false            || false
+        false               | false             | true              | true          || false            || false
+        false               | true              | false             | true          || false            || true
+        false               | true              | true              | false         || false            || true
+        false               | false             | false             | true          || true             || false
+        false               | false             | true              | false         || true             || false
+        false               | true              | false             | false         || false            || true
+        false               | false             | false             | false         || true             || false
```

Comment:
You know I dislike conditionals in expectations for tests :)

If I'm looking at this right, shouldFail is only true when scanForTestClasses=false and hasTestClasses=true, I think this is actually just testing existing behavior where we don't find tests when we think we should. It doesn't matter what we do with test definitions.

I also don't think we care about scanForTestClasses.  They don't impact how we select from test definitions.

Any test that has `addTestDefsDir` as false isn't exercising the test definition changes either.

Maybe part of what you're trying to do is mixed up in applying `java-library` because that comes with certain defaults. I think we only care about checking how skipped tests work with and without test definitions.

Reply:
I still find this table really clear and useful.  With the changes to remove properties and changes to what happens on non-existent tests, there is now some duplication here, yes.  But I'd rather be comprehensive and have this fully describe all combinations with their intended results and have it catch any regressions.

I don't think there's harm in a little redundancy here, I'd rather leave it.